datafusion/src/datasource/object_store.rs

```diff
@@ -0,0 +1,108 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+//! Object Store abstracts access to an underlying file/object storage.
+
+use crate::datasource::local::LocalFileSystem;
+use crate::error::Result;
+use std::any::Any;
+use std::collections::HashMap;
+use std::fmt::Debug;
+use std::io::Read;
+use std::sync::{Arc, RwLock};
+
+/// Objct Reader for one file in a object store
+pub trait ObjectReader {
```

Comment:
I think the traits in this file should be `async` (or at least offer an async version) -- as all the major object store implementations I know of in Rust are `async` and it seems that is the way the ecosystem is headed.

At the very least I would imagine `async fn get_reader(...) -> Box<dyn Read>` which would allow it to make a non blocking async network call to fetch the bytes. 

I am not sure if the thing returned (`Box<dyn Read>`) would benefit from being something async as well (aka an async stream) to allow streaming reads without necessarily buffering the entire `start/length` range.  At the moment I am inclined to keep it as `Box<dyn Read>`) for simplicity